File: leetcode/03047-find-the-largest-area-of-square-inside-two-rectangles/solution.cpp
class Solution {
public:
    long long largestSquareArea(vector<vector<int>>& bottomLeft, vector<vector<int>>& topRight) {
        long long ans = 0;
        int n = bottomLeft.size();
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                int bottom_left = max(bottomLeft[i][0], bottomLeft[j][0]);
                int top_right = min(topRight[i][0], topRight[j][0]);
                if (bottom_left >= top_right)
                    continue;

                int bottom_right = max(bottomLeft[i][1], bottomLeft[j][1]);
                int top_left = min(topRight[i][1], topRight[j][1]);
                if (bottom_right >= top_left) 
                    continue;

                int width = min(top_right - bottom_left, top_left - bottom_right);
                ans = max(ans, 1LL * width * width);
            }
        }
        return ans;
    }
};

File: leetcode/03047-find-the-largest-area-of-square-inside-two-rectangles/solution.rs
use std::cmp::{max, min};

impl Solution {
    pub fn largest_square_area(bottom_left: Vec<Vec<i32>>, top_right: Vec<Vec<i32>>) -> i64 {
        let n = bottom_left.len();
        let mut ans = 0_i64;

        for i in 0..n {
            for j in i+1..n {
                let bl = max(bottom_left[i][0], bottom_left[j][0]);
                let tr = min(top_right[i][0], top_right[j][0]);
                if bl >= tr {
                    continue 
                }
                
                let br = max(bottom_left[i][1], bottom_left[j][1]);
                let tl = min(top_right[i][1], top_right[j][1]);
                if br >= tl {
                    continue 
                }

                let width = min(tr - bl, tl - br) as i64;
                ans = max(ans, width * width);
            }
        }

        ans 
    }
}